### PR TITLE
hyprland: 0.38.1 -> 0.39.1

### PR DESCRIPTION
**Pull request by fufexan**
## Description of changes

[CHANGELOG 0.39.0](https://github.com/hyprwm/Hyprland/releases/tag/v0.39.0)
[CHANGELOG 0.39.1](https://github.com/hyprwm/Hyprland/releases/tag/v0.39.1)

Note: wlroots is now a fork of the upstream, and is built statically.

Closes #304143.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by wineee** [comment]
wlroots-hyprland is built statically and only used in hyprland, why not just build from [submodule](https://github.com/hyprwm/Hyprland/tree/main/subprojects)?

**Comment by fufexan** [comment]
We can do that, sure.

**Comment by fufexan** [comment]
@wineee I've tried making it build the subprojects directly but the derivation fails at the linking step. Not sure what might be wrong, maybe someone has an idea.

```diff
diff --git a/pkgs/applications/window-managers/hyprwm/hyprland/default.nix b/pkgs/applications/window-managers/hyprwm/hyprland/default.nix
index 4d8747b8f66b..cc0379ce1684 100644
--- a/pkgs/applications/window-managers/hyprwm/hyprland/default.nix
+++ b/pkgs/applications/window-managers/hyprwm/hyprland/default.nix
@@ -3,6 +3,7 @@
 , fetchFromGitHub
 , pkg-config
 , makeWrapper
+, cmake
 , meson
 , ninja
 , binutils
@@ -23,11 +24,12 @@
 , pciutils
 , systemd
 , tomlplusplus
-, udis86-hyprland
+, udis86
+, python3
 , wayland
 , wayland-protocols
 , wayland-scanner
-, wlroots-hyprland
+, wlroots
 , xcbutilwm
 , xwayland
 , debug ? false
@@ -45,7 +47,7 @@ assert lib.assertMsg (!enableNvidiaPatches) "The option `enableNvidiaPatches` ha
 assert lib.assertMsg (!hidpiXWayland) "The option `hidpiXWayland` has been removed. Please refer https://wiki.hyprland.org/Configuring/XWayland";
 
 let
-  wlr = wlroots-hyprland.override { inherit enableXWayland; };
+  # wlr = wlroots-hyprland.override { inherit enableXWayland; };
 in
 stdenv.mkDerivation (finalAttrs: {
   pname = "hyprland" + lib.optionalString debug "-debug";
@@ -54,12 +56,13 @@ stdenv.mkDerivation (finalAttrs: {
     owner = "hyprwm";
     repo = finalAttrs.pname;
     rev = "v${finalAttrs.version}";
-    hash = "sha256-piBUBQsQ8R20/lX9GMJjwdntal6yfJZMVCIi21XCZ8Y=";
+    hash = "sha256-7L5rqQRYH2iyyP5g3IdXJSlATfgnKhuYMf65E48MVKw=";
+    fetchSubmodules = true;
   };
 
   patches = [
     # make meson use the provided dependencies instead of the git submodules
-    "${finalAttrs.src}/nix/patches/meson-build.patch"
+    # "${finalAttrs.src}/nix/patches/meson-build.patch"
   ];
 
   postPatch = ''
@@ -82,6 +85,8 @@ stdenv.mkDerivation (finalAttrs: {
     makeWrapper
     meson
     ninja
+    cmake
+    python3
     pkg-config
     wayland-scanner
   ];
@@ -93,7 +98,7 @@ stdenv.mkDerivation (finalAttrs: {
   ];
 
   buildInputs =
-    wlr.buildInputs ++ [
+    wlroots.buildInputs ++ udis86.buildInputs ++ [
       cairo
       git
       hyprcursor
@@ -104,13 +109,11 @@ stdenv.mkDerivation (finalAttrs: {
       libinput
       libxkbcommon
       mesa
-      udis86-hyprland
       wayland
       wayland-protocols
       pango
       pciutils
       tomlplusplus
-      wlr
     ]
     ++ lib.optionals stdenv.hostPlatform.isMusl [ libexecinfo ]
     ++ lib.optionals enableXWayland [ libxcb xcbutilwm xwayland ]
@@ -130,7 +133,6 @@ stdenv.mkDerivation (finalAttrs: {
   ];
 
   postInstall = ''
-    ln -s ${wlr}/include/wlr $dev/include/hyprland/wlroots
     ${lib.optionalString wrapRuntimeDeps ''
       wrapProgram $out/bin/Hyprland \
         --suffix PATH : ${lib.makeBinPath [binutils pciutils stdenv.cc]}
@@ -145,6 +147,6 @@ stdenv.mkDerivation (finalAttrs: {
     license = licenses.bsd3;
     maintainers = with maintainers; [ wozeparrot fufexan ];
     mainProgram = "Hyprland";
-    platforms = wlr.meta.platforms;
+    platforms = wlroots.meta.platforms;
   };
 })
```

File: pkgs/applications/window-managers/hyprwm/hyprland/default.nix [hunk 1]
```diff
@@ -49,13 +49,12 @@ let
...
   pname = "hyprland" + lib.optionalString debug "-debug";
-  version = "0.38.1";
-
+  version = "0.39.1";
   src = fetchFromGitHub {
     owner = "hyprwm";
     repo = finalAttrs.pname;
     rev = "v${finalAttrs.version}";
-    hash = "sha256-6y422rx8ScSkjR1dNYGYUxBmFewRYlCz9XZZ+XrVZng=";
+    hash = "sha256-piBUBQsQ8R20/lX9GMJjwdntal6yfJZMVCIi21XCZ8Y=";
```
Comment (donovanglover) [inline pkgs/applications/window-managers/hyprwm/hyprland/default.nix:57]:
```suggestion
    hash = "sha256-Urb/njWiHYUudXpmK8EKl9Z58esTIG0PxXw5LuM2r5g=";
```
Builds for me with this hash (also the same hash I'm using for my 0.39.1 overlay)

Comment (fufexan) [reply pkgs/applications/window-managers/hyprwm/hyprland/default.nix]: Updated, thanks for the heads up.

**Comment by leiserfg** [comment]
Sorry for the noise but, what is blocking this?

**Comment by fufexan** [comment]
@leiserfg @wineee suggested we build the submodules directly instead of patching the build and using overridden derivations. But that doesn't work, as can be seen in my previous comment.

I plan on using [CMake instead of Meson for the Nix build](https://github.com/hyprwm/Hyprland/pull/5667), so we can merge this PR as is currently, and figure out the details later.

**Pull request merged**